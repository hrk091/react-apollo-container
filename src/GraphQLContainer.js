// @flow

import React, { Component } from "react"
import type { NodeQueryProps } from "./types"

export default class GraphQLContainer<A, T> extends Component<NodeQueryProps<A, T>> {

  render() {
    const { typename, data, loading, error } = this.props

    if (error) {
      console.error(error)
      return (
        <>
          <h3>データ取得エラー</h3>
          <h5>エラーが発生しました。ページをリロードしてください。</h5>
          <h5>事象が改善しない場合は、管理者にお問い合わせください。</h5>
        </>
      )
    }
    if (loading) {
      return <h3>Loading...</h3>
    }
    if (!data || !data[typename]) {
      return <h5>データが見つかりません。</h5>
    }

    const { children: Children } = this.props
    return <Children data={data[typename]} />
  }
}
