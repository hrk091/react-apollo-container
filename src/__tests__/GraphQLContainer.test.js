import React from "react"
import { MockedProvider } from "@apollo/react-testing"
import { render } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"

import { useQuery } from "@apollo/react-hooks"
import gql from "graphql-tag"

import GraphQLContainer from "../GraphQLContainer"
import { waitForDomChange } from "@testing-library/dom"

describe("GraphQLContainer", () => {
  // Setup

  const mockQuery = gql`
    query Contact($id: ID!) {
      contact(id: $id) {
        name
      }
    }
  `
  const variables = {
    id: "dummyId"
  }
  const typename = "contact"
  const data = { name: "dummyName" }

  function DummyComponent() {
    const queryProps = useQuery(mockQuery, { variables })
    return (
      <GraphQLContainer typename={typename} {...queryProps}>
        {({ data }) => JSON.stringify(data)}
      </GraphQLContainer>
    )
  }

  // Tests

  it("renders loading state initially", () => {
    // Given
    const mocks = [
      {
        request: { query: mockQuery, variables },
        result: {
          data: {
            [typename]: data
          }
        }
      }
    ]
    const spy = jest.fn()
    function DummyComponent() {
      const queryProps = useQuery(mockQuery, { variables })
      return (
        <GraphQLContainer typename={typename} {...queryProps}>
          {({ data }) => {
            spy()
            return <div />
          }}
        </GraphQLContainer>
      )
    }

    // When
    const { container, debug } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent />
      </MockedProvider>
    )

    // Then
    expect(spy).toHaveBeenCalledTimes(0)
  })

  it("renders graphQL result without crashing", async () => {
    // Given
    const mocks = [
      {
        request: { query: mockQuery, variables },
        result: {
          data: {
            [typename]: data
          }
        }
      }
    ]
    function DummyComponent() {
      const queryProps = useQuery(mockQuery, { variables })
      return (
        <GraphQLContainer typename={typename} {...queryProps}>
          {({ data }) => JSON.stringify(data)}
        </GraphQLContainer>
      )
    }

    // When
    const { container, debug } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent />
      </MockedProvider>
    )
    await waitForDomChange({ container })

    // Then
    expect(container).toHaveTextContent(JSON.stringify(data))
  })

  it("renders error message without crashing when error occurred without crashing", async () => {
    // Given
    const mocks = [
      {
        request: { query: mockQuery, variables },
        error: new Error("Some Error")
      }
    ]

    // When
    const { container, debug } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent />
      </MockedProvider>
    )
    await waitForDomChange({ container })

    // Then
    expect(container).toMatchSnapshot()
  })

  it("renders message 'No Data' without crashing when query target not found", async () => {
    // Given
    const mocks = [
      {
        request: { query: mockQuery, variables },
        result: {
          data: {
            [typename]: undefined
          }
        }
      }
    ]

    // When
    const { container, debug } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent />
      </MockedProvider>
    )
    await waitForDomChange({ container })

    // Then
    expect(container).toMatchSnapshot()
  })

  it("renders message 'No Data' without crashing if typename is wrong", async () => {
    // Given
    const mocks = [
      {
        request: { query: mockQuery, variables },
        result: {
          data: {
            [typename]: data
          }
        }
      }
    ]

    function WrongTypenameComponent() {
      const queryProps = useQuery(mockQuery, { variables })
      return (
        <GraphQLContainer typename={"anotherTypeName"} {...queryProps}>
          {({ data }) => JSON.stringify(data)}
        </GraphQLContainer>
      )
    }

    // When
    const { container, debug } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <WrongTypenameComponent />
      </MockedProvider>
    )
    await waitForDomChange({ container })

    // Then
    expect(container).toMatchSnapshot()
  })
})
