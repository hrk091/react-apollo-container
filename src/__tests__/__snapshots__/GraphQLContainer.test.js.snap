// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLContainer renders error message without crashing when error occurred without crashing 1`] = `
<div>
  <h3>
    データ取得エラー
  </h3>
  <h5>
    エラーが発生しました。ページをリロードしてください。
  </h5>
  <h5>
    事象が改善しない場合は、管理者にお問い合わせください。
  </h5>
</div>
`;

exports[`GraphQLContainer renders message 'No Data' without crashing if typename is wrong 1`] = `
<div>
  <h5>
    データが見つかりません。
  </h5>
</div>
`;

exports[`GraphQLContainer renders message 'No Data' without crashing when query target not found 1`] = `
<div>
  <h5>
    データが見つかりません。
  </h5>
</div>
`;
